packages/CollapsibleCard/src/CollapsibleCard.js

```diff
@@ -1,83 +1,71 @@
-  const [labelTextId] = React.useState(() => `collapsible-card-label_${uuidv4()}`);
-  const [metadataId] = React.useState(() => `collapsible-card-metadata_${uuidv4()}`);
-  const {
-    "CollapsibleCard.Avatar": avatar,
-    "CollapsibleCard.Metadata": metadata,
-    children: otherElements,
-  } = extractChildren(children, ["CollapsibleCard.Avatar", "CollapsibleCard.Metadata"]);
-
-  React.useEffect(() => {
-    if (!isCollapsed && onExpand) {
-      onExpand();
-    }
-  }, [isCollapsed, onExpand]);
+  const { children, initialIsCollapsed, isEditing, isFirstRow, isLastRow, isMiddleRow, onToggleIsCollapsed } = props;
+  const [isCollapsed, setIsCollapsed] = React.useState(initialIsCollapsed);
 
-  function getLabel() {
-    return (
-      <sc.Label>
-        {avatar}
-        <div>
-          <sc.LabelText id={labelTextId}>{label}</sc.LabelText>
-          {metadata ? React.cloneElement(metadata, { id: metadataId }) : null}
-        </div>
-      </sc.Label>
-    );
+  function handleToggleIsCollapsed() {
+    onToggleIsCollapsed(!isCollapsed);
+    setIsCollapsed(oldIsCollapsed => !oldIsCollapsed);
   }
 
-  function handleClick() {
-    setIsCollapsed(prevIsCollapsed => !prevIsCollapsed);
-  }
+  const thingsToShare = {
+    handleToggleIsCollapsed,
+    isCollapsed,
+    isEditing,
+    isFirstRow,
+    isLastRow,
+    isMiddleRow,
+    onToggleIsCollapsed,
+  };
 
   return (
-    <sc.CollapsibleCard
-      iconAlign="right"
-      label={typeof label === "function" ? label({ isCollapsed }) : getLabel()}
-      isCollapsed={isCollapsed}
-      onClick={handleClick}
-      hasAvatar={!!avatar}
-      hasLabelOnly={!avatar && !metadata}
-      triggerAriaDescribedby={`${labelTextId} ${metadata ? metadataId : ""}`}
-      {...moreProps}
-    >
-      <sc.Content hasDivider={hasDivider}>{otherElements}</sc.Content>
-    </sc.CollapsibleCard>
+    <CollapsibleCardContext.Provider value={thingsToShare}>
+      <sc.CollapsibleCard
+        aria-expanded={!isCollapsed}
+        isCollapsed={isCollapsed}
+        isEditing={isEditing}
+        isFirstRow={isFirstRow}
+        isMiddleRow={isMiddleRow}
+        isLastRow={isLastRow}
+      >
+        {children}
+      </sc.CollapsibleCard>
+    </CollapsibleCardContext.Provider>
   );
 }
 
 const propTypes = {
   children: PropTypes.node,
-
-  /** If has a divider between collapsible header and content. */
-  hasDivider: PropTypes.bool,
-
-  /** Label text as the card title, can be a render function. */
-  label: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
-
-  /** Callback function when expand the card. */
-  onExpand: PropTypes.func,
+  initialIsCollapsed: PropTypes.bool,
+  isEditing: PropTypes.bool,
+  isFirstRow: PropTypes.bool,
+  isMiddleRow: PropTypes.bool,
+  isLastRow: PropTypes.bool,
+  onToggleIsCollapsed: PropTypes.func,
 };
 
 const defaultProps = {
   children: null,
-  hasDivider: false,
-  label: null,
-  onExpand: null,
+  initialIsCollapsed: true,
```

Comment:
It seems like this component can only be uncontrolled? Should it be more flexible? if there are some form errors inside the collapsible content, we might want to prevent them from closing until users fix the errors? Just some thoughts 😄

Reply:
I've updated so it can be either controlled or uncontrolled, and added appropriate stories.